core/ledger/kvledger/txmgmt/privacyenabledstate/namespace_helper.go

```diff
@@ -0,0 +1,121 @@
+	for _, ns := range deployedccInfoProvider.Namespaces() {
+		retNamespaces = append(retNamespaces, ns)
+	}
+
+	// add namespaces for implicit collections
+	mspids, err := channelInfoProvider.GetAllMSPIDs()
+	if err != nil {
+		return nil, err
+	}
+	implicitCollNames := deployedccInfoProvider.ImplicitCollectionNames(mspids)
+	for _, implicitCollName := range implicitCollNames {
+		retNamespaces = append(retNamespaces, derivePvtDataNs("_lifecycle", implicitCollName))
+		retNamespaces = append(retNamespaces, deriveHashedDataNs("_lifecycle", implicitCollName))
+	}
+
+	// add namespaces for chaincodes and explicit collections
+	chaincodesInfo, err := deployedccInfoProvider.AllChaincodesInfo(channelName, &simpleQueryExecutor{vdb})
+	if err != nil {
+		return nil, err
+	}
+	for _, ccInfo := range chaincodesInfo {
+		retNamespaces = append(retNamespaces, ccInfo.Name)
+		if ccInfo.ExplicitCollectionConfigPkg == nil {
+			continue
+		}
+		for _, config := range ccInfo.ExplicitCollectionConfigPkg.Config {
+			collConfig := config.GetStaticCollectionConfig()
+			if collConfig != nil {
+				retNamespaces = append(retNamespaces, derivePvtDataNs(ccInfo.Name, collConfig.Name))
+				retNamespaces = append(retNamespaces, deriveHashedDataNs(ccInfo.Name, collConfig.Name))
+			}
+		}
+	}
+
+	return retNamespaces, nil
+}
+
+// simpleQueryExecutor implements ledger.SimpleQueryExecutor interface
+type simpleQueryExecutor struct {
+	vdb statedb.VersionedDB
+}
+
+func (sqe *simpleQueryExecutor) GetState(ns string, key string) ([]byte, error) {
+	versionedValue, err := sqe.vdb.GetState(ns, key)
+	if err != nil {
+		return nil, err
+	}
+	var value []byte
+	if versionedValue != nil {
+		value = versionedValue.Value
+	}
+	return value, nil
+}
+
+func (sqe *simpleQueryExecutor) GetStateRangeScanIterator(ns string, startKey string, endKey string) (commonledger.ResultsIterator, error) {
+	dbItr, err := sqe.vdb.GetStateRangeScanIterator(ns, startKey, endKey)
+	if err != nil {
+		return nil, err
+	}
+	itr := &resultsItr{ns: ns, dbItr: dbItr}
+	return itr, nil
+}
+
+// GetPrivateDataHash is not implemented and should not be called
+func (sqe *simpleQueryExecutor) GetPrivateDataHash(namespace, collection, key string) ([]byte, error) {
+	return nil, errors.New("not implemented yet")
+}
+
+type resultsItr struct {
+	ns    string
+	dbItr statedb.ResultsIterator
+}
+
+// Next implements method in interface ledger.ResultsIterator
+func (itr *resultsItr) Next() (commonledger.QueryResult, error) {
+	queryResult, err := itr.dbItr.Next()
+	if err != nil {
+		return nil, err
+	}
+	// itr.updateRangeQueryInfo(queryResult)
+	if queryResult == nil {
+		return nil, nil
+	}
+	versionedKV := queryResult.(*statedb.VersionedKV)
+	return &queryresult.KV{Namespace: versionedKV.Namespace, Key: versionedKV.Key, Value: versionedKV.Value}, nil
+}
+
+// Close implements method in interface ledger.ResultsIterator
+func (itr *resultsItr) Close() {
+	itr.dbItr.Close()
+}
```

Comment:
```suggestion
// Next implements method in interface ledger.ResultsIterator
func (itr *resultsItr) Next() (commonledger.QueryResult, error) {
	queryResult, err := itr.dbItr.Next()
	return queryResult.(commonledger.QueryResult), err
}
```
Wouldn't this work? Let me know if I am missing something.

Again the redirection is forced due to duplication of the `QueryResult` interface. In the `statedb.go`, we have a `QueryResult` that is used by the each statedb implementation. Then, in the `common/ledger/ledger_interface.go`, we have the same interface. Due to the mixed usage, we couldn't simply reuse the iterator present at the `VersionedDB`.

Reply:
Since statecouchdb returns *statedb.VersionedKV but the caller expects *queryresult.KV, the data type conversion is needed.